transport-native-epoll/src/main/c/io_netty_channel_libaio_DirectFileDescriptorController.c

```diff
@@ -0,0 +1,347 @@
+
+    struct io_control * theControl = (struct io_control *) malloc(sizeof(struct io_control));
+
+    theControl->errorInfoClass = (*env)->FindClass(env, "io/netty/channel/libaio/ErrorInfo");
+    if (theControl->errorInfoClass == NULL)
+    {
+       return;
+    }
+
+
+    /// The ErrorInfoClass is barely used. The VM would crash in the event of an error without this GlobalRef
+    theControl->errorInfoClass = (jclass)(*env)->NewGlobalRef(env, (jobject)(theControl->errorInfoClass));
+
+    theControl->errorInfoConstr = (*env)->GetMethodID(env, theControl->errorInfoClass, "<init>", "(Ljava/lang/Object;ILjava/lang/String;)V");
+
+    if (theControl->errorInfoConstr == NULL)
+    {
+       return;
+    }
+
+    /// The ErrorInfoClass is barely used. The VM would crash in the event of an error without this GlobalRef
+    theControl->errorInfoConstr = (jmethodID)(*env)->NewGlobalRef(env, (jobject)(theControl->errorInfoConstr));
+
+    theControl->ioContext = libaioContext;
+    theControl->events = events;
+    theControl->iocb = iocb;
+    theControl->queueSize = queueSize;
+    theControl->iocbPut = 0;
+    theControl->iocbGet = 0;
+    theControl->used = 0;
+    pthread_mutex_init(&(theControl->iocbLock), 0);
+
+    jobject bufferContext = (*env)->NewDirectByteBuffer(env, theControl, sizeof(struct io_control));
+    return bufferContext;
+}
+
+JNIEXPORT void JNICALL Java_io_netty_channel_libaio_DirectFileDescriptorController_deleteContext(JNIEnv* env, jclass clazz, jobject contextPointer) {
+
+    int i;
+    struct io_control * theControl = getIOControl(env, contextPointer);
+    io_queue_release(theControl->ioContext);
+
+    // Releasing each individual iocb
+    for (i = 0; i < theControl->queueSize; i++)
+    {
+       free(theControl->iocb[i]);
+    }
+
+    (*env)->DeleteGlobalRef(env, (jobject)(theControl->errorInfoClass));
+    (*env)->DeleteGlobalRef(env, (jobject)(theControl->errorInfoConstr));
+
+    free(theControl->iocb);
+
+    free(theControl->events);
+
+    free(theControl);
+}
+
+JNIEXPORT void JNICALL Java_io_netty_channel_libaio_DirectFileDescriptorController_close(JNIEnv* env, jclass clazz, jint fd) {
+   if (close(fd) < 0) {
+       throwIOException(env, exceptionMessage("Error closing file:", errno));
+   }
+}
+
+JNIEXPORT int JNICALL Java_io_netty_channel_libaio_DirectFileDescriptorController_open(JNIEnv* env, jclass clazz, jstring path) {
+
+    const char* f_path = (*env)->GetStringUTFChars(env, path, 0);
+
+    int res = open(f_path, O_RDWR | O_CREAT | O_DIRECT, 0666);
+
+    (*env)->ReleaseStringUTFChars(env, path, f_path);
+
+    if (res < 0) {
+       throwIOException(env, exceptionMessage("Cannot open file:", errno));
+    }
+
+    return res;
+}
+
+static inline void submit(JNIEnv * env, io_context_t ioContext,  struct iocb * iocb) {
+    int result = io_submit(ioContext, 1, &iocb);
+
+    if (result < 0) {
+        if (result == -EAGAIN) {
```

Comment:
It's being called on the callback. The Object should live between the submission and the callback,
otherwise you could have a VM crash if the user didn't hold the reference himself.